### PR TITLE
Modelselection

### PR DESCRIPTION
**Pull request by statrita2004**
One model-selection rule using Random Forest classification of Pudlo 2016 implemented.

### CODE REVIEW

File: abcpy/modelselections.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        """To be overwritten by any sub-class: returns a model chosen by the modelselection
+        procedure most suitable to the obersved data set observations. It is assumed that observations is a 
+        list of n same type elements(eg., The observations can be a list containing n timeseries, n 
+        graphs or n np.ndarray). Further two optional integer arguments n_samples and n_samples_per_param
+        is supplied denoting the number of samples in the refernce table and the data points in each 
+        simulated data set.
+        
+        Parameters
+        ----------
+        data: python list
```
Comment (mschoengens) [inline abcpy/modelselections.py]: There is no parameter data..

File: abcpy/modelselections.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        self.backend = backend
+        self.rng = np.random.RandomState(seed)
+        self.reference_table_calculated = None
+        
+
+        raise NotImplemented
+        
+        
+    @abstractmethod
+    def modelchoice(self, observations, n_samples = 1000, n_samplee_per_param = 100):
```
Comment (mschoengens) [inline abcpy/modelselections.py]: n_samplee_per_param; -e +s

File: abcpy/modelselections.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        """To be overwritten by any sub-class: returns a model chosen by the modelselection
+        procedure most suitable to the obersved data set observations. It is assumed that observations is a 
+        list of n same type elements(eg., The observations can be a list containing n timeseries, n 
+        graphs or n np.ndarray). Further two optional integer arguments n_samples and n_samples_per_param
+        is supplied denoting the number of samples in the refernce table and the data points in each
```
Comment (mschoengens) [inline abcpy/modelselections.py]: Is the parameter `observations` actually the same as in the inference schemes sample method? In the sample method, observations can also be a set of n numbers. Is this ok here as well?

File: abcpy/modelselections.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+                if self.reference_table_models[ind1] == self.model_array[ind2]:
+                    label[ind1] = ind2 
+
+        # Define the classifier 
+        classifier = RandomForestClassifier(n_estimators = N_tree, max_features=int(n_try_fraction*self.reference_table_statistics.shape[1]), bootstrap=True, random_state=self.seed)
+        classifier.fit(self.reference_table_statistics, label)            
+
+        return(self.model_array[int(classifier.predict(self.statistics_calc.statistics(observations)))])
+
+    def posteriorprobability(self, observations, n_samples = 1000, n_samples_per_param = 1, N_tree = 100, n_try_fraction = 0.5):
```
Comment (mschoengens) [inline abcpy/modelselections.py]: See comment above on  N_tree and n_try_fraction.

File: abcpy/modelselections.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        -------
+        abcpy.models.Model
+            A model which are of type abcpy.models.Model
+            
+        """
+        
+        raise NotImplemented
+
+    @abstractmethod
+    def posteriorprobability(self, observations):
```
Comment (mschoengens) [inline abcpy/modelselections.py]: The method would be more readable if we use posterior_probability. What do you think?

File: abcpy/modelselections.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        for ind in range(len(self.reference_table_models)):
+            pred_error[ind] = 1 - classifier.predict_proba(self.statistics_calc.statistics(self.reference_table_data[ind]))[0][int(label[ind])] 
+
+        # Estimate a regression function with prediction error as response on summary statitistics of the reference table                
+        regressor = RandomForestRegressor(n_estimators = N_tree)
+        regressor.fit(self.reference_table_statistics,pred_error)
+
+        return(1-regressor.predict(self.statistics_calc.statistics(observations)))
+        
+class _RemoteContextgeneratereferencetable:
```
Comment (mschoengens) [inline abcpy/modelselections.py]: Please use CamelCase to increase readability: _RemoteContextGenerateReferenceTable, maybe  _RemoteContextForReferenceTable is even clearer?

File: abcpy/modelselections.py (hunk 1 above)
Comment (statrita2004) [inline abcpy/modelselections.py]: Changed it to observations

File: abcpy/modelselections.py (hunk 2 above)
Comment (statrita2004) [inline abcpy/modelselections.py]: 1. changed to select_model
2. corrected

File: abcpy/modelselections.py (hunk 5 above)
Comment (statrita2004) [inline abcpy/modelselections.py]: changed to posterior_probability

File: abcpy/modelselections.py (hunk 6 above)
Comment (statrita2004) [inline abcpy/modelselections.py]: changed to _RemoteContextForReferenceTable

File: abcpy/modelselections.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    """
+    def __init__(self, model_array, statistics_calc, backend, seed = None):
+        self.model_array = model_array
+        self.statistics_calc = statistics_calc
+        self.backend = backend
+        self.rng = np.random.RandomState(seed)
+        self.seed = seed 
+        self.reference_table_calculated = 0
+
+    def modelchoice(self, observations, n_samples = 1000, n_samples_per_param = 1, N_tree = 100, n_try_fraction = 0.5):
```
Comment (statrita2004) [inline abcpy/modelselections.py]: agreed and moved them to _init_

File: abcpy/modelselections.py (hunk 4 above)
Comment (statrita2004) [inline abcpy/modelselections.py]: agreed and moved them to _init_

File: abcpy/modelselections.py (hunk 3 above)
Comment (statrita2004) [inline abcpy/modelselections.py]: Yes. It's same as inference. I follow the same convention.

File: abcpy/modelselections.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
+from abc import ABCMeta, abstractmethod
+import numpy as np
+from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
+
+
+
+class ModelSelections(metaclass = ABCMeta):
+    """This abstract base class defines a model selection rulr of how to choose model from a set of models
```
Comment (statrita2004) [inline abcpy/modelselections.py]: It's difficult to explain the high level goal. In general this should be known to a practitioner who wants to perform modelselection.

Comment (statrita2004) [inline abcpy/modelselections.py]: Fixed also abcpy.rst. 

But notice there is some warning regarding uniform _init_ in master. I don't understand that, please check that after merging.

**Pull request merged**